### PR TITLE
AP_HAL_Linux: support UNIX FIFO(named pipe) interface

### PR DESCRIPTION
**Pull request by rsaxvc**
Create a new serial device backend that connects to a pair(RX/TX) of named pipes.

### CODE REVIEW

File: libraries/AP_HAL_Linux/UARTDriver.cpp [hunk 1]
```diff
@@ -185,17 +191,31 @@ AP_HAL::OwnPtr<SerialDevice> UARTDriver::_parseDevicePath(const char *arg)
...
             }
-            device = new UDPDevice(_ip, _base_port, false, true);
+            device = new UDPDevice(_ip, base_port, false, true);
 
         }
+    } else if(strcmp(protocol,"fifo") == 0) {
+        const char * rx_fifo = _ip;
+        const char * tx_fifo = _port;
+        if(stat(_ip, &st) != 0) {
+                AP_HAL::panic("stat() RX fifo %s failed", rx_fifo);
```
Comment (rsaxvc) [inline libraries/AP_HAL_Linux/UARTDriver.cpp:206]: Indentation seems off

File: libraries/AP_HAL_Linux/UARTDevice.cpp [hunk 2]
```diff
@@ -179,8 +179,7 @@ void UARTDevice::set_parity(int v)
         } else {
             t.c_cflag &= ~PARODD;
         }
-    }
-    else {
+    } else {
```
Comment (rsaxvc) [inline libraries/AP_HAL_Linux/UARTDevice.cpp:182]: I can undo this change to UARTDevice.cpp.

**Comment by lucasdemarchi** [comment]
> Create a new serial device backend that connects to a pair(RX/TX) of named pipes.

I don't see the motivation anywhere. What's the use case here?

**Comment by lucasdemarchi** [comment]
Overall looks good. I left some comments and would like to have a justification for this support.

**Comment by rsaxvc** [comment]
> I don't see the motivation anywhere. What's the use case here?

I use this to do simple I/O logging with tee. I also use it to provide a connection like a virtual serial port with flow control to connect to another program on the same device. I'm using OpenHD for forwarding MAVLink over raw(IP-less) 802.11 frames.

**Comment by rsaxvc** [comment]
I'm sorry about that, I accidentally pushed an old branch. I've addressed your comments and am rebuilding to test now.

**Comment by rsaxvc** [comment]
This is the correct patch and works well. However, I think I've found something I'll need to address. If the pipe is closed, ardupilot will have one thread always busy failing to access them. This might take me a bit to get to. Let me know if you prefer I close this and reopen later.

**Comment by rsaxvc** [comment]
OpenHD added more fully-featured MavLink networking, so this is no longer needed.

**Pull request closed without merging**